src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PairHMMLikelihoodCalculationEngine.java

```diff
@@ -0,0 +1,396 @@
+
+            applyPCRErrorModel(readBases, readInsQuals, readDelQuals);
+            capMinimumReadQualities(read, readQuals, readInsQuals, readDelQuals);
+
+            // Create a new copy of the read and sets its base qualities to the modified versions.
+            result.add(createQualityModifiedRead(read, readBases, readQuals, readInsQuals, readDelQuals));
+        }
+        return result;
+    }
+
+    private static void capMinimumReadQualities(final GATKRead read, final byte[] readQuals, final byte[] readInsQuals, final byte[] readDelQuals) {
+        for( int i = 0; i < readQuals.length; i++ ) {
+            readQuals[i] = (byte) Math.min(0xff & readQuals[i], read.getMappingQuality()); // cap base quality by mapping quality, as in UG
+            readQuals[i] =    setToFixedValueIfTooLow( readQuals[i],    PairHMM.BASE_QUALITY_SCORE_THRESHOLD,  QualityUtils.MIN_USABLE_Q_SCORE );
+            readInsQuals[i] = setToFixedValueIfTooLow( readInsQuals[i], QualityUtils.MIN_USABLE_Q_SCORE,       QualityUtils.MIN_USABLE_Q_SCORE );
+            readDelQuals[i] = setToFixedValueIfTooLow( readDelQuals[i], QualityUtils.MIN_USABLE_Q_SCORE,       QualityUtils.MIN_USABLE_Q_SCORE );
+        }
+    }
+
+    private static byte setToFixedValueIfTooLow(final byte currentVal, final byte minQual, final byte fixedQual){
+        return currentVal < minQual ? fixedQual : currentVal;
+    }
+
+    private static Map<GATKRead, byte[]> buildGapContinuationPenalties(final List<GATKRead> reads, final byte gapPenalty) {
+        return reads.stream().collect(Collectors.toMap(read -> read,
+                                                       read -> Utils.dupBytes(gapPenalty, read.getLength())));
+    }
+
+    private void writeDebugLikelihoods(final LikelihoodMatrix<Haplotype> likelihoods) {
+        if (!writeLikelihoodsToFile || likelihoodsStream == null) {
+            return;
+        }
+
+        final List<GATKRead> reads = likelihoods.reads();
+        final List<Haplotype> haplotypes = likelihoods.alleles();
+        for (int i = 0; i < reads.size(); i++) {
+            for (int j = 0; j < haplotypes.size(); j++) {
+                writeDebugLikelihoods(reads.get(i), haplotypes.get(j), likelihoods.get(j, i));
+            }
+        }
+        likelihoodsStream.flush();
+    }
+
+    private void writeDebugLikelihoods(final GATKRead processedRead, final Haplotype haplotype, final double log10l){
+        likelihoodsStream.printf("%s %s %s %s %s %s %f%n",
+                haplotype.getBaseString(),
+                new String(processedRead.getBases()),
+                SAMUtils.phredToFastq(processedRead.getBaseQualities()),
+                SAMUtils.phredToFastq(ReadUtils.getBaseInsertionQualities(processedRead)),
+                SAMUtils.phredToFastq(ReadUtils.getBaseDeletionQualities(processedRead)),
+                SAMUtils.phredToFastq(constantGCP),
+                log10l);
```

Comment:
because we got fooled but this code recently, can you add a comment here to say that the precision of log10l like only 5 or 6 digits (ie not all digits are printed)?